sympy/stats/stochastic_process_types.py

```diff
@@ -489,22 +494,72 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
+                for idx in idx2state):
+                return S.Zero # a RandomIndexedSymbol cannot go to different states simultaneously
             i, result = -1, 1
+            conds = And(*[Intersection(idx2state[idx].as_set(), state_space).as_relational(idx) for idx in idx2state])
+            if not isinstance(conds, And):
+                return self.probability(conds, given_condition & TransitionMatrixOf(self, trans_probs)
+                                                & StochasticStateSpaceOf(self, state_space))
+            conds = conds.args
             while i > -len(conds):
                 result *= self.probability(conds[i], conds[i-1] & \
                             TransitionMatrixOf(self, trans_probs) & \
                             StochasticStateSpaceOf(self, state_space))
                 i -= 1
             if isinstance(given_condition, (TransitionMatrixOf, StochasticStateSpaceOf)):
                 return result * Probability(conds[i])
+            if isinstance(given_condition, And):
+                idx_sym = conds[i].atoms(RandomIndexedSymbol)
+                prob, count = S(0), 0
+                for gc in given_condition.args:
+                    if gc.atoms(RandomIndexedSymbol) == idx_sym:
+                        prob += gc.rhs if isinstance(gc.lhs, Probability) else gc.lhs
+                        count += 1
+                if isinstance(state_space, FiniteSet) and \
+                    count == len(state_space) - 1:
+                    given_condition = Eq(Probability(conds[i]), S(1) - prob)
             if isinstance(given_condition, Eq):
                 if not isinstance(given_condition.lhs, Probability) or \
                     given_condition.lhs.args[0] != conds[i]:
                     raise ValueError("Probability for %s needed", conds[i])
                 return result * given_condition.rhs
 
+        if isinstance(condition, Or):
+            conds, prob_sum = condition.args, S(0)
+            idx2state = dict()
+            for cond in conds:
+                idx, state = (cond.lhs, cond.rhs) if isinstance(cond.lhs, RandomIndexedSymbol) else \
+                                (cond.rhs, cond.lhs)
+                idx2state[idx] = cond if idx2state.get(idx, None) is None else \
+                                           idx2state[idx] | cond
+            conds = Or(*[Intersection(idx2state[idx].as_set(), state_space).as_relational(idx)
+                        for idx in idx2state])
+            if not isinstance(conds, Or):
+                return self.probability(conds, given_condition & TransitionMatrixOf(self, trans_probs)
+                                                & StochasticStateSpaceOf(self, state_space))
+            conds = conds.args
+            for cond in conds:
+                prob_sum += self.probability(cond, given_condition &
+                            TransitionMatrixOf(self, trans_probs) &
+                            StochasticStateSpaceOf(self, state_space))
```

Comment:
There same to be many constructions of `given_condition & TransitionMatrixOf(self, trans_probs) & StochasticStateSpaceOf(self, state_space)` in different branches. Perhaps that could be stored in a variable (say `cond`), and then this could be reduced to:
```py
prob_sum = sum([self.probability(c, cond) for c in conds.args])
```

Reply:
Yes, right, I will do that cleaning in the next commit.